### ISSUE TITLE
Visual UI Editor

### ISSUE THREAD

**Issue by rsredsq** [AtomicGameEngine/AtomicGameEngine#138]
Atomic Editor needs a Visual UI Editor. Editing *.tb files manually is kinda terrible horror :D

**Comment by JoshEngebretson** [comment]
This is quite true, though with some practice it can be a lot quicker than setting up UI hierarchy in script.  One thing I would like to do is change the names of the files to *.ui.txt as this makes it more clear what they are...

... another complication for the UI Editor, is that TurboBadger has some global state for skinning and is designed to have one skin loaded at a time.  This means that editing skinned/themed game ui's in the editor (which also uses TB) isn't possible without some work on the TB core, or by possibly having the editor spawn a separate rendering process

A first step would be to just have a simple UI editor resource frame, or maybe a ui edit mode, maybe a first step is actually to look around for some inspiration on how a UI editor should integrate with the editor :)

**Comment by rsredsq** [comment]
Separate process for UI Editor is better. Would be good, if UI Editor could launch separately like a tool(e.g. TileEditor)

**Comment by rsredsq** [comment]
It's not really kinda of must have things, editor could live without that yet, but to provide a full editing system for users(developers) Atomic Editor should have it

**Comment by JoshEngebretson** [comment]
There are quite a few examples in the TB repo:

https://github.com/fruxo/turbobadger/tree/master/Demo/demo01/ui_resources

There is even a little UI editor of sorts, the *.tb.txt format is pretty powerful and succinct, while JSON would be a better serialization target, a parser would need to be written and some of the constructs might be difficult to reproduce.  If writing a UI by hand, I think the tb format wins and the JSON would end up being more verbose and less elegant.  For a visual editor, JSON probably wins.

BTW, it is worth building the TurboBadger demo from the repo and running it.

**Comment by rsredsq** [comment]
Parsing JSON file with TurboBadger data should be pretty easy. Write get/set method for each TB Widget attribute, in UiWidget (they are about [100](https://github.com/fruxo/turbobadger/blob/master/resources/metadata/tb_widgets_attr.tb.txt).)
JSON format is very popular and easy for manually editing(in my opinion), so no need for learning new format. 

Probably creating JSON parser is a good way to have alternative for TB format.
